Add method to invite user to Trip

Vielheim opened the pull request:
- Add InviteUserView, which is embedded inside EditTripView
- On invite, the view model will fetch the User, then fetch current Trip copy
- Reason for copy is that we don't want to save the changes the user did in the EditTripView
- Maybe the user doesn't want to save the changes and just want to add a new user
- So, we will let the user explicitly decide when he/she wants to save
- Finally, update the trip copy and save into Firebase

Corner cases:
- User doesn't exist
- User already added

Commit:
Add method to invite user to Trip

- Add InviteUserView, which is embedded inside EditTripView
- On invite, the viewmodel will fetch the User, then fetch Trip copy
- Reason for copy is that we don't want to save the changes the user did in the EditTripView
- Maybe the user doesn't want to save the changes and just want to add a new user
- Let the user explicitly do the saving
- Next, update the copy and save into Firebase

tanruiquan commented on TourMate/TourMate/Frontend/User/Controllers/FirebaseUserController.swift:
```diff
@@ -46,4 +46,27 @@ struct FirebaseUserController: UserController {
+        }
+
+        // should not have more than 1 user with the same email
+        guard adaptedUsers.count <= 1 else {
+            preconditionFailure()
+        }
```
Although this should not happen, `preconditionFailure()` would kill the app so I'm thinking we should do something safer?
```suggestion
        guard adaptedUsers.count <= 1 else {
            assertionFailure()
            return (nil, "More than one user found")
        }
```

Vielheim replied:
Ah ok, I was following `getUser()` to be consistent. Lemme update both functions.

Vielheim replied:
> Do you think we should include a list view of all the attending users in `TripVIew`?

Okie, I'll add it in as well

tanruiquan commented on TourMate/TourMate/Frontend/User/Controllers/FirebaseUserController.swift:
```diff
@@ -46,4 +46,27 @@ struct FirebaseUserController: UserController {
+            preconditionFailure()
+        }
+
+        guard let adaptedUser = adaptedUsers.first as? FirebaseAdaptedUser else {
+            return (nil, "") // user not found
+        }
```
I think we can pass in the errorMessage?
```suggestion
        guard let adaptedUser = adaptedUsers.first as? FirebaseAdaptedUser else {
            return (nil, "User not found") // user not found
        }
```

Vielheim replied:
If we pass in the errorMessage, the front-end will register this as an error. Then it will display the error page instead. It's like that time we mentioned like (nil, "") means no item found in db?

tanruiquan replied:
Right, I forgot about that 👌

tanruiquan reviewed (commented):
Do you think we should include a list view of all the attending users in `TripVIew`?

Commit:
Update error checks

Commit:
Merge branch 'main' into invite-users

# Conflicts:
#	TourMate/TourMate/Frontend/Trip/Views/EditTripView/EditTripView.swift

Commit:
Add AttendeesView

Vielheim commented:
Added AttendeesView
- Added `attendees` as a @Published property in TripViewModel
- Added method in TripViewModel to fetch attendees
- Will be fetched when the view refreshes

Added UserIconView
- Displays the User's image and name

Made `fetchUser` in UserController and FirebaseUserController more general
- Able to fetch a user based on a `field` and `value` that the field is supposed to be equal to.
- Would recommend just using it for keys (email, id etc.)

tanruiquan reviewed (approved):
LGTM 👍

tanruiquan reviewed (approved):
LGTM 👍
